nav2_safety_nodes/src/nav2_safety_node.cpp

```diff
@@ -0,0 +1,229 @@
+  }
+  timer_->cancel();
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+SafetyZone::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(logger_, "Cleaning up");
+  safety_polygon_pub_.reset();
+  speed_limit_pub_.reset();
+  timer_.reset();
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+SafetyZone::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(logger_, "Shutting down");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void
+SafetyZone::laserCallback(
+  const sensor_msgs::msg::LaserScan::SharedPtr message)
+{
+  // project the laser into a point cloud
+  sensor_msgs::msg::PointCloud2 cloud;
+
+  try {
+    projector_.projectLaser(*message, cloud);
+  } catch (...) {
+    RCLCPP_WARN(logger_, "Unable to project laser into cloud");
+    return;
+  }
+
+  // Transform the cloud if necessary
+  if (!base_frame_.empty() && cloud.header.frame_id != base_frame_) {
+    try {
+      cloud = tf2_->transform(cloud, base_frame_, tf2::durationFromSec(tf_tolerance_));
+      sensor_data_.push(cloud);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_WARN(logger_, "Transform failure: %s", ex.what());
+      return;
+    }
+  } else if (cloud.header.frame_id == base_frame_) {
+    sensor_data_.push(cloud);
+  }
+}
+
+double
+SafetyZone::dotProduct(
+  const Eigen::Vector3d & pt1,
+  const Eigen::Vector3d & pt2)
+{
+  return pt1[0] * pt2[1] - pt1[1] * pt2[0];
+}
+
+// Function for detecting if cloud points are inside safety polygon or not
+int
+SafetyZone::detectPoints(
+  const sensor_msgs::msg::PointCloud2 & cloud,
+  const std::vector<geometry_msgs::msg::Point> & safety_zone)
+{
+  int num_points_in_zone = 0;
+  sensor_msgs::PointCloud2ConstIterator<float> iter_x(cloud, "x");
+  sensor_msgs::PointCloud2ConstIterator<float> iter_y(cloud, "y");
+  sensor_msgs::PointCloud2ConstIterator<float> iter_z(cloud, "z");
+  // iterating through cloud points
+  for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z) {
+    double px = *iter_x, py = *iter_y, pz = *iter_z;
+    int count_same_side_results = 0, on_edge = 0;
+    // iterating through polygon points
+    for (int i = 0; i < int(safety_zone.size()); ++i) {
+      geometry_msgs::msg::Point vertex = safety_zone[i];
+      geometry_msgs::msg::Point next_vertex = safety_zone[(i + 1) % safety_zone.size()];
+      Eigen::Vector3d affine_segment = {next_vertex.x - vertex.x, next_vertex.y - vertex.y, next_vertex.z - vertex.z};
+      Eigen::Vector3d affine_point = {px - vertex.x, py - vertex.y, pz - vertex.z};
+      double dot = dotProduct(affine_segment, affine_point);
+      if (dot > 0) {
+        count_same_side_results++;
+      } else if (dot < 0) {
+        count_same_side_results--;
+      } else {
+        on_edge++;
+      }
+    }
+    // checking if point is on same side of all edges
+    if (abs(count_same_side_results == int(safety_zone.size())) || on_edge == 2) {
```

Comment:
This is wrong. Can you see why its wrong?

Reply:
Yes, My mistake